packages/components/src/navigator/navigator-provider/component.tsx

```diff
@@ -40,6 +69,49 @@ function UnconnectedNavigatorProvider(
 			path: initialPath,
 		},
 	] );
+	const [ screens, dispatch ] = useReducer( screensReducer, [] );
+	const [ matchedPath, setMatchedPath ] = useState< MatchedPath >( false );
+	const currentMatch = useRef< MatchedPath >( false );
+
+	useEffect( () => {
+		const updateMatch = ( newMatch: MatchedPath ) => {
+			setMatchedPath( newMatch );
+			currentMatch.current = newMatch;
+		};
+
+		const resolvePath = ( path: string ) => {
+			const newMatch = patternMatch( path, screens );
+			if (
+				currentMatch.current !== false &&
+				newMatch !== false &&
+				isShallowEqual( newMatch, currentMatch.current )
```

Comment:
If let's say we add a new screen causing us to check if the matched screen changes, we do a shallow comparison to avoid re-rendering components if the matching function returned the same output.

Reply:
Ah, got it! Thank you for explaining (and adding an inline comment too!)